src/plugins/intel_gpu/src/graph/primitive_inst.cpp

```diff
@@ -507,26 +522,19 @@ event::ptr primitive_inst::realloc_if_needed() {
         }
     }
 
-    if (_node->is_type<gather>() || _node->is_type<permute>()) {
-        // For the nodes which can be optimized at runtime, input memory is used as output memory
-        // So there is no need to reallocate output memory
-        if (can_be_optimized())
+    // Clear out memory if if was previously reused, but now primitive can't be optimized
+    if (_node->is_type<gather>() || _node->is_type<permute>() || _node->is_type<reshape>() || _node->is_type<reorder>()) {
```

Comment:
Probably we should remove these type checks and logic below should be generalized for all primitive types. Though it can be done separately

Reply:
I will apply it in the next PR